<h1>План исследования<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-1">Цель исследования</a></span></li><li><span><a href="#Шаг-1.-Исследуем-таблицы" data-toc-modified-id="Шаг-1.-Исследуем-таблицы-2">Шаг 1. Исследуем таблицы</a></span><ul class="toc-item"><li><span><a href="#1.1.-Подключение-к-базе-данных" data-toc-modified-id="1.1.-Подключение-к-базе-данных-2.1">1.1. Подключение к базе данных</a></span></li><li><span><a href="#1.2.-Исследуем-таблицы:-books" data-toc-modified-id="1.2.-Исследуем-таблицы:-books-2.2">1.2. Исследуем таблицы: <em>books</em></a></span></li><li><span><a href="#1.3.-Исследуем-таблицы:-authors" data-toc-modified-id="1.3.-Исследуем-таблицы:-authors-2.3">1.3. Исследуем таблицы: <em>authors</em></a></span></li><li><span><a href="#1.4.--Исследуем-таблицы:-publishers" data-toc-modified-id="1.4.--Исследуем-таблицы:-publishers-2.4">1.4.  Исследуем таблицы: <em>publishers</em></a></span></li><li><span><a href="#1.5.--Исследуем-таблицы:-ratings" data-toc-modified-id="1.5.--Исследуем-таблицы:-ratings-2.5">1.5.  Исследуем таблицы: <em>ratings</em></a></span></li><li><span><a href="#1.6.--Исследуем-таблицы:-reviews" data-toc-modified-id="1.6.--Исследуем-таблицы:-reviews-2.6">1.6.  Исследуем таблицы: <em>reviews</em></a></span></li></ul></li><li><span><a href="#Шаг-2.-Отвечаем-на-вопросы-исследования" data-toc-modified-id="Шаг-2.-Отвечаем-на-вопросы-исследования-3">Шаг 2. Отвечаем на вопросы исследования</a></span><ul class="toc-item"><li><span><a href="#2.1.-Cколько-книг-вышло-после-1-января-2000-года?" data-toc-modified-id="2.1.-Cколько-книг-вышло-после-1-января-2000-года?-3.1">2.1. Cколько книг вышло после 1 января 2000 года?</a></span></li><li><span><a href="#2.2.-Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="2.2.-Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку.-3.2">2.2. Для каждой книги посчитать количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#2.3.-Определить-издательство,-которое-выпустило-наибольшее-число-книг(за-исключением-брошюр)" data-toc-modified-id="2.3.-Определить-издательство,-которое-выпустило-наибольшее-число-книг(за-исключением-брошюр)-3.3">2.3. Определить издательство, которое выпустило наибольшее число книг(за исключением брошюр)</a></span></li><li><span><a href="#2.4.-Oпределить-автора-с-самой-высокой-средней-оценкой-книг-(учитываем-только-книги-с-50-и-более-оценками)" data-toc-modified-id="2.4.-Oпределить-автора-с-самой-высокой-средней-оценкой-книг-(учитываем-только-книги-с-50-и-более-оценками)-3.4">2.4. Oпределить автора с самой высокой средней оценкой книг (учитываем только книги с 50 и более оценками)</a></span></li><li><span><a href="#2.5.-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="2.5.-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5">2.5. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Шаг-3.-Общий-вывод" data-toc-modified-id="Шаг-3.-Общий-вывод-4">Шаг 3. Общий вывод</a></span></li></ul></div>

# Анализ базы данных

## Цель исследования

Необходимо сформулировать ценностное предложение для нового продукта(сервиса дл чтения книг по подписке).
В нашем распоряжении база данных, в которой есть информация о книгах, издательствах, авторах, а также пользовательские обзоры книг - нужно ее проанализировать и ответить на следующие вопросы:

* *сколько книг вышло после 1 января 2000 года;*

* *для каждой книги посчитать количество обзоров и среднюю оценку;*

* *определить издательство, которое выпустило наибольшее число книг(за исключением брошюр);*

* *определить автора с самой высокой средней оценкой книг (учитываем только книги с 50 и более оценками);*

* *посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок*.

## Шаг 1. Исследуем таблицы

### 1.1. Подключение к базе данных

Импортируем библиотеки:

In [1]:
import pandas as pd 
from sqlalchemy import create_engine

Устанавливаем параметры:

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432, # портподключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],  
                                                         db_config['pwd'],  
                                                         db_config['host'],  
                                                         db_config['port'],  
                                                         db_config['db'])

Cохраняем коннектор:

In [3]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Напишем функцию для выполнения SQL-запросов:

In [4]:
def sql_func(query):
    return pd.read_sql(query, con = engine)

### 1.2. Исследуем таблицы: *books*

In [5]:
query_books = '''SELECT 
                    *
                 FROM 
                    books
                 LIMIT 5;
                        '''
sql_func(query_books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
query_books_info = '''SELECT 
                    *
                 FROM 
                    books;
                        '''
sql_func(query_books_info).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Таблица **books** содержит данные о книгах. Состоит из *1000 строк*, пропуски отсутствуют. Содержит *6 столбцов* со следующей информацией:

* *book_id* — идентификатор книги;

* *author_id* — идентификатор автора;

* *title* — название книги;

* *num_pages* — количество страниц;

* *publication_date* — дата публикации книги;

* *publisher_id* — идентификатор издателя.

### 1.3. Исследуем таблицы: *authors*

In [7]:
query_authors = '''SELECT 
                    *
                 FROM 
                    authors
                 LIMIT 5;
                        '''
sql_func(query_authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
query_authors_info = '''SELECT 
                    *
                 FROM 
                    authors;
                        '''
sql_func(query_authors_info).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Таблица **authors** содержит данные об авторах. Состоит из *636 строк*, пропуски отсутствуют. Содержит *2 столбца* со следующей информацией:

* *author_id* — идентификатор автора;

* *author* — имя автора.

### 1.4.  Исследуем таблицы: *publishers*

In [9]:
query_publishers = '''SELECT 
                    *
                 FROM 
                    publishers
                 LIMIT 5;
                        '''
sql_func(query_publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
query_publishers_info = '''SELECT 
                    *
                 FROM 
                    publishers;
                        '''
sql_func(query_publishers_info).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица **publishers** содержит данные об  издательствах. Состоит из *340 строк*, пропуски отсутствуют. Содержит *2 столбца* со следующей информацией:

* *publisher_id* — идентификатор издательства;

* *publisher* — название издательства.

### 1.5.  Исследуем таблицы: *ratings*

In [11]:
query_ratings = '''SELECT 
                    *
                   FROM 
                    ratings
                 LIMIT 5;
                        '''
sql_func(query_ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
query_ratings_info = '''SELECT 
                    *
                   FROM 
                    ratings;
                        '''
sql_func(query_ratings_info).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица **ratings** содержит данные о пользовательских оценках книг. Состоит из *6456 строк*, пропуски отсутствуют. Содержит *4 столбца* со следующей информацией:

* *rating_id* — идентификатор оценки;

* *book_id* — идентификатор книги;

* *username* — имя пользователя, оставившего оценку;

* *rating* — оценка книги.


### 1.6.  Исследуем таблицы: *reviews*

In [13]:
query_reviews = '''SELECT 
                    *
                   FROM 
                    reviews
                 LIMIT 5;
                        '''
sql_func(query_reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
query_reviews_info = '''SELECT 
                    *
                   FROM 
                    reviews;
                        '''
sql_func(query_reviews_info).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Таблица **ratings** содержит данные о пользовательских обзорах на книги. Состоит из *2793 строк*, пропуски отсутствуют. Содержит *4 столбца* со следующей информацией:

* *review_id* — идентификатор обзора;

* *book_id* — идентификатор книги;

* *username* — имя пользователя, написавшего обзор;

* *text* — текст обзора.

## Шаг 2. Отвечаем на вопросы исследования

### 2.1. Cколько книг вышло после 1 января 2000 года?

In [15]:
query_count_books = '''SELECT
                            COUNT(book_id) AS book_cnt
                       FROM
                            books
                       WHERE
                            publication_date > '2000-01-01';
                            '''
sql_func(query_count_books)

book_cnt
0       819

После *1 января 2000 года* вышло **819 книг**.

### 2.2. Для каждой книги посчитать количество обзоров и среднюю оценку.

In [16]:
query_reviews_rating = '''SELECT
                    books.book_id AS book_id,
                    books.title AS title,
                    review_cnt,
                    ROUND(average_rating,2) AS average_rating
                FROM
                    books
                    LEFT JOIN 
                        (SELECT
                            book_id,
                            COUNT(review_id) AS review_cnt
                        FROM
                            reviews
                        GROUP BY
                            book_id
                        ) AS reviews_1 ON books.book_id = reviews_1.book_id
                    LEFT JOIN
                        (SELECT
                            book_id,
                            AVG(rating) AS average_rating
                        FROM
                            ratings
                        GROUP BY
                            book_id
                        ) AS ratings_1 ON books.book_id = ratings_1.book_id

'''

sql_func(query_reviews_rating)

book_id                                              title  review_cnt  \
0        652          The Body in the Library (Miss Marple  #3)         2.0   
1        273                                          Galápagos         2.0   
2         51                           A Tree Grows in Brooklyn         5.0   
3        951  Undaunted Courage: The Pioneering First Missio...         2.0   
4        839                                        The Prophet         4.0   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland         4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)         2.0   
997      148                                          Christine         3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)         2.0   
999      828                           The Plot Against America         2.0   

     average_rating  
0              4.50  
1              4.50  
2              4.25  
3              4.00  
4              4.29  
..              ...  
995            4.23  
996            5.00  
997            3.43  
998            3.50  
999            3.00  

[1000 rows x 4 columns]

### 2.3. Определить издательство, которое выпустило наибольшее число книг(за исключением брошюр)

In [17]:
query_more_50_pages_books = '''SELECT
                    publishers.publisher AS publisher,
                    COUNT(DISTINCT books.title) AS more_50_pages_books
                FROM
                    books
                    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
                WHERE
                    books.num_pages > 50

                GROUP BY
                    publishers.publisher
                ORDER BY
                    more_50_pages_books DESC
                LIMIT 1;
                
                
                    

'''

sql_func(query_more_50_pages_books)

publisher  more_50_pages_books
0  Penguin Books                   42

Издательство, выпустившее наибольшее число книг(за исключением брошюр), называется ***Penguin Books***. Выпустили *42 книги*.

### 2.4. Oпределить автора с самой высокой средней оценкой книг (учитываем только книги с 50 и более оценками)

In [18]:
query_best_author = '''SELECT
                    books.author_id AS author_id,
                    authors.author AS author,
                    ROUND(AVG(SBQ.book_rating_avg),3) AS author_rating_avg
                  FROM 
                    books
                    LEFT JOIN authors ON authors.author_id = books.author_id
                    INNER JOIN (
                            SELECT
                                book_id,
                                AVG(rating) AS book_rating_avg
                            FROM
                                ratings
                            GROUP BY
                                book_id
                            HAVING 
                                COUNT(*) >= 50
                                ) AS SBQ on SBQ.book_id = books.book_id
                   GROUP BY
                      books.author_id,
                      authors.author
                   ORDER BY
                      author_rating_avg DESC
                    LIMIT 1
;


'''

sql_func(query_best_author)

author_id                      author  author_rating_avg
0        236  J.K. Rowling/Mary GrandPré              4.284

Автором с самой высокой оценкой книг оказалась ***Джоан Роулинг*** (*средний рейтинг - 4.28*).

### 2.5. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [19]:
query_test = '''SELECT 
                    ROUND(AVG(cnt_reviews)) AS avg_reviews
                FROM (
                        SELECT 
                            reviews.username, 
                            COUNT(text) AS cnt_reviews
                        FROM 
                            reviews
                            JOIN (
                                    SELECT 
                                        username, 
                                        COUNT(rating) 
                                    FROM 
                                        ratings
                                    GROUP BY 
                                        username
                                    HAVING COUNT(rating) > 50) AS query ON query.username=reviews.username
                        GROUP BY 
                            reviews.username) AS query_2;
                            ; 
                    
'''

sql_func(query_test)

avg_reviews
0         24.0

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24 обзора**.

## Шаг 3. Общий вывод

С помощью базы данных нам было необходимо найти ответы на ряд вопросов. Мы воспользовались *SQL - запросами*.

Ниже приведем ответы на поставленные вопросы:

1. *Cколько книг вышло после 1 января 2000 года?*

После *1 января 2000 года* вышло **819 книг**.

2. *Для каждой книги посчитать количество обзоров и среднюю оценку*

In [20]:
sql_func(query_reviews_rating)

book_id                                              title  review_cnt  \
0        652          The Body in the Library (Miss Marple  #3)         2.0   
1        273                                          Galápagos         2.0   
2         51                           A Tree Grows in Brooklyn         5.0   
3        951  Undaunted Courage: The Pioneering First Missio...         2.0   
4        839                                        The Prophet         4.0   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland         4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)         2.0   
997      148                                          Christine         3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)         2.0   
999      828                           The Plot Against America         2.0   

     average_rating  
0              4.50  
1              4.50  
2              4.25  
3              4.00  
4              4.29  
..              ...  
995            4.23  
996            5.00  
997            3.43  
998            3.50  
999            3.00  

[1000 rows x 4 columns]

3. *Какое издательство выпустило наибольшее число книг(за исключением брошюр)?*

Издательство, выпустившее наибольшее число книг(за исключением брошюр), называется ***Penguin Books***. Выпустили *42 книги*.

4. *Определить автора с самой высокой средней оценкой книг (учитываем только книги с 50 и более оценками).*

Автором с самой высокой оценкой книг оказалась ***Джоан Роулинг*** (*средний рейтинг - 4.28*).

5. *Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.*

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24 обзора**.